File: 540-single-element-in-a-sorted-array/single-element-in-a-sorted-array.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        lo=0
        hi=len(nums)-1
        if len(nums)==1:
            return nums[0]
        if nums[lo]!=nums[lo+1]:
            return nums[lo]
        if nums[hi]!=nums[hi-1]:
            return nums[hi]
        
        while lo<=hi:
            mid=lo+(hi-lo)//2
            if nums[mid]!=nums[mid-1] and nums[mid]!=nums[mid+1]:
                return nums[mid]
            elif nums[mid]==nums[mid-1]:
                if mid&1==0:hi=mid-1
                else:lo=mid+1
            elif nums[mid]==nums[mid+1]:
                if mid&1==0:lo=mid+1
                else:hi=mid-1
        
        
        

File: 540-single-element-in-a-sorted-array/single-element-in-a-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNonDuplicate = function(nums) {
    let lo=0;
    let hi=nums.length-1
    while(lo<=hi){
        let mid=Math.floor((lo+hi)/2)
        if(nums[mid]!==nums[mid-1] && nums[mid]!==nums[mid+1]){
            return nums[mid]
        }else if (mid%2==0){
            if(nums[mid]===nums[mid-1]){
                hi=mid-1
            }else{
                lo=mid+1
            }
        }else if(mid%2!==0){
            if(nums[mid]===nums[mid-1]){
                lo=mid+1
            }else{
                hi=mid-1
            }
        }
    }
    
};